src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectSystem/RazorProjectHostBase.cs

```diff
@@ -249,6 +279,130 @@ Task IProjectDynamicLoadComponent.UnloadAsync()
             return DisposeAsync();
         }
 
+        private void ProjectManager_Changed(object sender, ProjectChangeEventArgs args)
+        {
+            switch (args.Kind)
+            {
+                case ProjectChangeKind.DocumentRemoved:
+                case ProjectChangeKind.DocumentAdded:
+                case ProjectChangeKind.DocumentChanged:
+                case ProjectChangeKind.ProjectChanged:
+                    // These changes can come in bursts so we don't want to overload the publishing system. Therefore,
+                    // we enqueue publishes and then publish the latest project after a delay.
+
+                    EnqueuePublish(args.Newer);
+                    break;
+                case ProjectChangeKind.ProjectAdded:
+                    Publish(args.Newer);
+                    break;
+                case ProjectChangeKind.ProjectRemoved:
+                    RemovePublishing(args.Older);
+                    break;
+            }
+        }
+
+        internal void EnqueuePublish(ProjectSnapshot projectSnapshot)
+        {
+            // A race is not possible here because we use the main thread to synchronize the updates
+            // by capturing the sync context.
+
+            CommonServices.ThreadingService.VerifyOnUIThread();
```

Comment:
When you split this logic into a different class you'll want use the JoinableTaskFactory method of verifying you're on the UI thread to ensure methods are unit testable.